### PR TITLE
Build and publish CUDA wheels

### PR DESCRIPTION
**Pull request by shoumikhin**
## The problem

The wheel can carry the CUDA delegate, but no published wheel contains one: there is no CUDA
row in any workflow, so a GPU user has to build from source.

## The change

Add the workflows that build and publish CUDA wheels for Linux x86_64 and aarch64, and a smoke test
that checks each wheel from the artifact itself. The aarch64 rows are validated on machines with no
GPU, so for those the smoke test cannot execute a model. It verifies the CUDA libraries are present,
that the declared runtime matches the wheel's CUDA version, that nothing resolves through the build
machine's toolkit, and that the shipped device code covers every GPU architecture the row claims.

```
executorch-1.5.0-cp312-cp312-manylinux_2_28_x86_64.whl    +cu130
```

A release publishes CUDA 12.6, 13.0 and 13.2, for Python 3.10 through 3.13. That is twelve rows in
each of the two workflows, so twenty-four wheels. A pull request builds one row per workflow, two
wheels, because a full matrix costs hours for little extra signal.

Which GPU architectures each row compiles for is chosen per row rather than detected on the builder.
Detecting it would produce a wheel carrying device code for whatever machine happened to build it,
which installs fine and then fails at the first GPU call.

The aarch64 CUDA 12.6 row also compiles for compute capability 8.7, which is an embedded module.
Every other row lists only the architectures the published PyTorch build for that train covers, and
by that rule 8.7 would be left out, because the generic aarch64 build of this train carries 8.0 and
9.0 only. It is included because this is the only row whose CUDA major version matches what that
module's software release ships, and because this wheel declares no PyTorch dependency: a user there
supplies the build that carries their architecture. Leaving 8.7 out does not protect them from a bad
pairing, it only removes the device code they need. Without it, a model reaching one of the shipped
optional operators, quantized matrix multiply, sort or random number generation, fails at the first
launch on that device.

Two guards keep a release honest:

- if the shared matrix generator stops offering a combination this policy advertises, the step fails
  instead of quietly publishing fewer wheels. A missing job is otherwise a green check for a wheel
  that was never built.
- if a row reaches the architecture list with no CUDA version, the build refuses rather than falling
  back to the builder's GPU. A TORCH_CUDA_ARCH_LIST that holds only named GPU families PyTorch
  accepts, such as "Hopper", is left to torch instead: those entries are dropped with a message and
  CMAKE_CUDA_ARCHITECTURES stays unset, because torch has already turned them into gencode flags by
  the time this runs and failing here would break a configuration that builds today. The three
  named forms CMake itself understands, "native", "all" and "all-major", are rejected before this
  logic runs:
  torch resolves the list with its own bundled CUDA architecture module, which does not know
  those names and stops the configure. That is upstream behaviour, not something this change
  introduces or can work around, so a caller has to name architectures explicitly.

Windows CUDA is deliberately absent. The separate shared libraries this wheel exists to ship are
Linux only today, so a Windows CUDA wheel would carry a delegate a C++ application still could not
link.

Both workflows cancel an in-flight run when a newer one starts for the same ref, matching what the
CPU wheel workflows do. A ciflow tag re-push moves the tag rather than creating a new ref, so
without this the superseded build keeps a set of GPU builders busy for the rest of its run.

## Test plan

- built the full release matrix and confirmed each wheel's contents match the row it claims: the
  CUDA libraries present, the CUDA runtime declared, and device code for every GPU architecture the
  row advertises.
- ran a GPU model end to end from a CI-built wheel on three NVIDIA GPUs covering three device
  architectures, with output identical to eager PyTorch on each (largest absolute difference 0),
  and inspected the wheel for a fourth device it cannot execute on.
- ran the matrix filter over generated inputs, including incomplete and malformed ones, and
  confirmed it refuses rather than publishing a partial release: a missing CUDA version, a missing
  python, or a python present on rows this policy does not build are each reported by name.
- confirmed a CPU row still produces a CPU wheel on a builder that happens to have a CUDA toolkit
  installed.
- the newest architecture also ships in its portable form, so a GPU newer than any in the row can
  still run by having the driver compile it at load time. Checked with `cuobjdump --list-ptx`, since
  `--list-elf` prints identical output whether or not the portable form is present.
- every library that carries GPU device code covers the whole row on its own.
- the declared CUDA packages are compared against the expected set in BOTH directions. A one-way
  comparison accepted a wheel that omitted required packages, and a name-suffix comparison accepted
  cross-train names because for CUDA 13 the suffix is empty.
- the python axis is an allowlist, matching the CUDA axis. Testing only the disabled list let any
  python not on it through: a 3.9 row was emitted successfully.
- `install_utils.py` is in both CUDA workflows' path filters. It owns the supported CUDA train list
  and the toolkit detection, so a change there previously ran no CUDA wheel job.
- the headers the wheel ships as its C++ SDK are in both filters too, for the same reason and
  matching the CPU workflows. `extension/cuda` is already covered by the CUDA-specific entry, but
  the rest of the SDK is not, and a CUDA wheel publishes all of it.
- requesting the JetPack rows fails with its own reason instead of the generic empty-matrix message,
  since both of its lists are deliberately empty and no workflow asks for them.
- torchao keeps its CUDA channel where that channel exists. Falling back to its CPU channel was
  needed only on aarch64, where the CUDA channel publishes nothing, and doing it everywhere
  changed which torchao an x86_64 install resolves.
- the CUDA smoke test now asserts the QnnBackend and OpenvinoBackend registrations that a CPU Linux
  row asserts. The CUDA build enables OpenVINO on every Linux architecture and downloads the QNN
  SDK on x86_64, so a CUDA wheel carries both backends; a previous premise that "a CUDA row is not
  built with them" was false, and dropping the checks meant those two backends were unverified on
  every CUDA wheel.

The x86_64 rows land on a GPU runner, so the smoke test there exports a model through the delegate,
runs it and compares against eager. The aarch64 rows have no accelerator, so that one check reports
a skip naming the reason and the row still verifies the shipped libraries, their declared
dependencies, their loader paths and the device code they carry.

### CODE REVIEW

**Comment by shoumikhin** [comment]
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* #21816
* #21771
* #21770
* __->__ #21668

File: .github/workflows/build-wheels-cuda-aarch64-linux.yml [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+      - v[0-9]+.[0-9]+.[0-9]+-rc[0-9]+
+      - ciflow/binaries/*
+  workflow_dispatch:
+
+jobs:
+  generate-matrix:
+    uses: pytorch/test-infra/.github/workflows/generate_binary_build_matrix.yml@main
+    with:
+      package-type: wheel
+      os: linux-aarch64
```
Comment (digantdesai) [inline .github/workflows/build-wheels-cuda-aarch64-linux.yml:45]: do we have CI machines to test arm + cuda?

File: .ci/scripts/wheel/test_cuda_linux.py [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+    with tempfile.TemporaryDirectory() as work_dir:
+        test_shared_libraries.run_tests(Path(work_dir))
+    with tempfile.TemporaryDirectory() as work_dir:
+        test_cpp_sdk.run_tests(Path(work_dir))
+
+    test_base.run_tests(
+        model_tests=[
+            test_base.ModelTest(
+                model=Model.Mv3,
+                backend=Backend.XnnpackQuantizationDelegation,
```
Comment (digantdesai) [inline .ci/scripts/wheel/test_cuda_linux.py:524]: why not AOTI-CUDA?

Comment (digantdesai) [reply .ci/scripts/wheel/test_cuda_linux.py:524]: should we also make sure the AoT side of things also install the right version of torch when someone does 'pip install executorch' which gets these cuda libs, they better work with AoT torch cuda wheel generated AOTI blobs which uses cubin and other cuda runtime APIs.

**Review by digantdesai — APPROVED** [review]
stamping to unblock, left some comments.

File: .github/workflows/build-wheels-cuda-aarch64-linux.yml (hunk 1 above)
Comment (shoumikhin) [reply .github/workflows/build-wheels-cuda-aarch64-linux.yml:45]: Not for running on a GPU. These aarch64 runners have no accelerator, so CI builds the wheel and smoke-tests the import, but cannot execute a CUDA model.

I covered that gap by hand instead: I installed these wheels on a Thor (sm_110) and an Orin Nano (sm_87) and ran models there. Worth knowing the limit is the runner fleet, not the wheel.

File: .ci/scripts/wheel/test_cuda_linux.py (hunk 2 above)
Comment (shoumikhin) [reply .ci/scripts/wheel/test_cuda_linux.py:524]: Both fair points. On the naming, the test targets the CUDA delegate, which is the AOTI-CUDA path, so I can rename it to say that plainly if you prefer.

On the version match, that is the real risk you are pointing at: the AOTI blob is produced by the torch that exported the model, so it has to agree with the CUDA libraries in the wheel. Today the wheel declares the CUDA runtime as a pip dependency but does not pin the torch used for export. I would rather fix that as its own change than fold it in here, since it affects how we publish, not just this test.
